lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee

```diff
@@ -0,0 +1,141 @@
+define((require, exports, module) ->
+  'use strict'
+  oop = require('../../lib/oop')
+  TextHighlightRules = require('../../mode/text_highlight_rules').TextHighlightRules
+
+  PapeeriaLatexHighlightRules = ->
+    ###*
+    * We maintain a stack of nested LaTeX semantic types (e.g. "document", "section", "list"
+    * to be able to provide context for autocompletion and other functions. 
+    * Stack is constructed by the background highlighter;
+    * its elements are then propagated to * the editor session and become 
+    * available through getContext method.
+
+    * @param {destination} string
+    * @return {function} function, which correctly puts new type(destination) on stack
+   ###
+    pushState = (destination) ->
+      return (currentState, stack) ->
+        if currentState == 'start'
+          stack.push(currentState, destination)
+        else
+          stack.push(destination)
+        return destination
+
+    popState = (currentState, stack) ->
+      return stack.pop() or 'start'
```

Comment:
any sense to use LaTeX terminology in the state names? E.g. "document" as the root state